### ISSUE TITLE
✅ 🐞 Bug Report: Responsive css styles are not generated after the first call to get_builder_content_for_display - [ED-10594]

### ISSUE THREAD

**Issue by VadymPogorelov** [elementor/elementor#20555]
### Prerequisites

- [X] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [X] The issue still exists against the latest stable version of Elementor.

### Description

If you make two calls to the \Elementor\Plugin::instance()->frontend->get_builder_content_for_display() method, then after editing and saving the second post, styles for tablet and mobile are not generated.

### Steps to reproduce

1. Activating the Twenty Twenty One Theme and Elementor
2. Create two pages using Elementor, in which to insert the Spacer widget and set different heights for Desktop, Tablet and Mobile: https://prnt.sc/f5mJNUsZUldO
3. Let's insert the layouts of these pages into the header.php template (let's assume this is a custom header): https://prnt.sc/7zzvA-xPHyFh
I have pages with id 105 and 7
```
echo \Elementor\Plugin::instance()->frontend->get_builder_content_for_display( 105 );
echo \Elementor\Plugin::instance()->frontend->get_builder_content_for_display( 7 );
```
4. On first view of the page in frontend (immediately after creating pages 105 and 7 and calling them in header.php) everything is fine: https://prnt.sc/VGTU0o9CgKtO
5. Now let's make changes to the second page (with id 7) and save it: https://prnt.sc/-G2h34YTj1w5
6. At this point, the Elementor removes the previously generated styles and regenerates them the first time the page is accessed. And we only get the style for desktop. There are no styles for tablet and mobile: 
https://prnt.sc/L5Iz2tAzq03p and https://prnt.sc/W82YkTZnG0_A

### Isolating the problem

- [X] This bug happens with only Elementor plugin active (and Elementor Pro).
- [X] This bug happens with a Blank WordPress theme active ([Hello theme](https://wordpress.org/themes/hello-elementor/)).
- [X] I can reproduce this bug consistently using the steps above.

### System Info

<details>
<summary>System Info</summary>

````txt


					
== Server Environment ==
	Operating System: WINNT
	Software: Apache
	MySQL version: mariadb.org binary distribution v10.8.4
	PHP Version: 7.4.30
	PHP Memory Limit: 1536M
	PHP Max Input Vars: 4000
	PHP Max Post Size: 50M
	GD Installed: Yes
	ZIP Installed: Yes
	Write Permissions: All right
	Elementor Library: Connected

== WordPress Environment ==
	Version: 6.1.1
	Site URL: http://test/tmp
	Home URL: http://test/tmp
	WP Multisite: No
	Max Upload Size: 50 MB
	Memory limit: 40M
	Max Memory limit: 1536M
	Permalink Structure: /%year%/%monthnum%/%day%/%postname%/
	Language: en-US
	Timezone: 0
	Admin Email: [REDACTED]
	Debug Mode: Inactive

== Theme ==
	Name: Twenty Twenty-One
	Version: 1.7
	Author: the WordPress team
	Child Theme: No

== User ==
	Role: administrator
	WP Profile lang: en_US
	User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:107.0) Gecko/20100101 Firefox/107.0

== Active Plugins ==
	Elementor
		Version: 3.8.1
		Author: Elementor.com


== Elementor Experiments ==
	Optimized DOM Output: Active by default
	Improved Asset Loading: Active by default
	Improved CSS Loading: Active by default
	Inline Font Icons: Inactive by default
	Accessibility Improvements: Active by default
	Additional Custom Breakpoints: Active by default
	Import Export Website Kit: Active by default
	Hide native WordPress widgets from search results: Active by default
	admin_menu_rearrangement: Inactive by default
	Flexbox Container: Inactive by default
	Landing Pages: Active by default
	Color Sampler: Active by default
	Favorite Widgets: Active by default
	Admin Top Bar: Active by default


== Log ==
	


== Elementor - Compatibility Tag ==
	

				

````

</details>

**Comment by nicholaszein** [comment]
Hello @VadymPogorelov! 

Thank you for submitting your issue! 🙏 

Please allow us some time to investigate it, and get back to you with our findings. 🧐 

Best regards

**Comment by hein-obox** [comment]
@VadymPogorelov 
Do you still have this issue?
Do you also have it on other installations?
Does the effect change if you switch the additional breakpoints experiment on or off?

I have tested it with various Elementor Core versions, but I can't reproduce this problem. 
I have tested it with various experiments on and off, but I still I can't reproduce this problem.

**Comment by VadymPogorelov** [comment]
Just did all the steps again:

1) A "clean" installation of WordPress https://prnt.sc/KhrBXf-yw5_O

2) The "Twenty Twenty One" theme is active https://prnt.sc/rFBwdEXrVik6

3) The "Elementor" plugin is active https://prnt.sc/iJoyz71dCU1l No settings are changed!

4) Two pages with a widget "Spacer" are created. 
For each widget a parameter "Size" was set up with values: 
100px for Desktop
80px for Tablet
60px for Mobile
A border with "2px solid red" is specified for each Spacer.
First page with id=6 https://prnt.sc/t2gfxi80WjQT
Second page with id=11 https://prnt.sc/e8SCBMbbXEBS

5) Insert the layouts of these pages in header.php: https://prnt.sc/v3YJW3rW2Wwj
(in the real theme these pages are used as custom header and footer and pasted in the appropriate places)

6) Now let's open any page, for example, Sample Page: https://prnt.sc/pQnuDUcw1hkq
The first time you visit the page, everything is fine (at all resolutions). For example, for tablet: https://prnt.sc/7YLuAvyZ279C

7) We look at the generated css file for the second page (with id=11). Because By default, Elementor stores custom styles for each post separately https://prnt.sc/jO-TVhSUm1kA , then open the corresponding file: https://prnt.sc/Iwm30LDi_wsm

8) Now let's change something on the page with id=11 and save it. (Important! It is the second of the inserted pages that needs to be changed. In this case, the first should not be edited.) For example, let's set the height of Spacer for tablet to 81px: https://prnt.sc/8QxzdjwNg9ou

9) When saving a post, Elementor removes the custom styles file so that it can be regenerated the first time it is accessed: https://prnt.sc/K2fwbUScTR_n

10) Let's update the tab with the Sample Page and see that the second Spacer at Tablet resolution has a size of 100px instead of the specified 81px: https://prnt.sc/M0oJbIEEfPSa

11) At the same time, the file with styles for the page with id=11 does not contain styles for tablet and mobile: https://prnt.sc/tOyuVGoBVSpi

Once again, I draw your attention: no Elementor settings have changed after installation!

**Comment by hein-obox** [comment]
@VadymPogorelov 

1. First question: does it make a difference if the Active Breakpoints experiment is on or off?
2. Does the styling get fixed when you click on Elementor > Tools > Regenerate CSS?

I have noticed a similar problem with Elementor templates which are used in a shortcode.

**Comment by hein-obox** [comment]
It would help to know these answers.

@VadymPogorelov

First question: does it make a difference if the Active Breakpoints experiment is on or off?
Does the styling get fixed when you click on Elementor > Tools > Regenerate CSS?

**Comment by hein-obox** [comment]
@nicholaszein 

I believe that this issue is solved in the latest update 3.10.0-cloud.
I tested it on 3.9.1 as well. The problem still occurs in that version. 

@VadymPogorelov 

Can you please check if this issue is solved once 3.9.2 or 3.10.0 come out?

**Comment by VadymPogorelov** [comment]
Issue not resolved after update 3.9.2: 
https://prnt.sc/yDwwXQa-HSsJ
https://prnt.sc/TYyQzF3aaD5Q

**Comment by hogash** [comment]
I also want to jump in and show how it also happens with `[elementor-template id="99"]` shortcodes. 

I recorded a video here https://d.pr/v/FoHXR8 . 

What i did to replicate these steps:
- created 2 templates, each containing headings;
- in the 2nd template, the heading has 60px on desktop, while on mobile 22px;
- i pasted both elementor template shortcodes inside a page;
- edit the 2nd template (BBB) and change the mobile font size;
- refresh the page where the 2 shortcodes are added, and on mobile it will show the 2nd template (BBB) without any mobile styles. In fact no mobile styles are added to this template's stylesheets. 

Only Elementor and Elementor PRO are installed, theme is Hello. Additional Custom Breakpoints is active.

Thanks!

**Comment by hogash** [comment]
Hi @nicholaszein @hein-obox 

Just tested with Elementor 3.10.0-beta3 and still not working eg: https://d.pr/v/glQ20a (same setup as my previous comment). The mobile styles of the 2nd template is not generating at all.

**Comment by black-eye** [comment]
Is it possible that this issue is related to https://github.com/elementor/elementor/issues/19394 ?

**Comment by hogash** [comment]
> Is it possible that this issue is related to #19394 ?

Could be. Both are related to mobile styles not being generated.

**Comment by VadymPogorelov** [comment]
I checked: if 'duplication_mode' is set to 'on' https://prnt.sc/eRJ0DSBb5xTy - responsive css is regenerated correctly after changing the second post (template) on the page.

**Comment by VadymPogorelov** [comment]
By default, 'duplication mode' is off and when css is regenerated after updating the second layout, the get_active_controls() method "does not see" breakpoint controls and, accordingly, css rules are not created for them

**Comment by schroedermatthias** [comment]
I was facing this issue as well:

1. Clearing Elementors CSS cache via Elementor -> Tools -> Regenerate CSS & Data
2. Calling page **A** in the frontend (responsive view) -> Page is fine
3. Calling page **B** in the frontend (responsive view) -> Page is broken

And the other way around:
1. Clearing Elementors CSS cache via Elementor -> Tools -> Regenerate CSS & Data
2. Calling page **B** in the frontend (responsive view) -> Page is fine
3. Calling page **A** in the frontend (responsive view) -> Page is broken

For me, setting "Additional Custom Breakpoints" under Elementor -> Settings -> Experiments to "Inactive" solved this bug as a workaround.


<img width="373" alt="Bildschirm­foto 2023-01-12 um 14 17 43" src="https://user-images.githubusercontent.com/3320701/212076285-290c1952-254e-4261-abfd-1b1090a291ea.png">

**Comment by VadymPogorelov** [comment]
Issue is not resolved after update 3.10.1.

Please note that since the "Custom breakpoints" experiment is enabled by default, responsiveness issues occur for everyone who uses custom header and footer based on Elementor templates on the page.

**Comment by tahmidbintaslim** [comment]
I also see this issue in the latest Elementor Version 3.10.2. After editing pages, CSS is not generated until manually go to Tool and Regenerate CSS

**Comment by VadymPogorelov** [comment]
Issue is not resolved after update 3.11.2

**Comment by pajtai** [comment]
We're having this issue too. @VadymPogorelov thanks for the steps to repro!

**Comment by pajtai** [comment]
@schroedermatthias thanks for the suggestion of disable "custom breakpoints" - that fixed it for us!

**Comment by VadymPogorelov** [comment]
Issue is not resolved after update 3.12.1
Edit a second page https://prnt.sc/HtAGlLxJ4cNA
Check in frontend https://prnt.sc/maaNUF8FeLA_

**Comment by VadymPogorelov** [comment]
Issue is not resolved after update 3.14

**Comment by SilverFoxA** [comment]
+ Issue not yet resolved

**Comment by nicholaszein** [comment]
Hello there!

We have great news! 🙌 

### 📢 We're happy to announce that the issue you raised was resolved in Elementor Core v3.15.0! 🥳 

✅ Feel free to check it out and update your plugin to the new version!

> _Check the changelog for more information:_
> [Changelog of Elementor and Elementor Pro](https://elementor.com/help/changelog-of-elementor/)

Cheers 🥂

**Comment by nicholaszein** [comment]
Hello everyone,

we have an update on this issue. 

Unfortunately, due to some unforeseen problems caused by the patch for this issue introduced in version **3.15.0**,  we had to [revert the fix](https://wordpress.org/plugins/elementor/#developers) on Elementor Core **v3.15.3**. 

In some cases, it was not possible to edit any document that contained a **Template Widget** (with a template selected), or a **Template Shortcode**. It also affected editing some **Theme Builder Site Parts**, causing the Editor to be stuck in the loading screen (#23408).

### We are currently working on an alternative fix for this issue and will be releasing a new fix as soon as possible. 💪 

We understand this might be frustrating, and we deeply apologize for the inconvenience. 😔 🙏 

Please bear with us as we find a new solution to this issue. I will be updating you immediately as I have any news. So if you don't see a comment from me, it means we are still working on a solution.

If you experience any problem with responsive styles not loading, we suggest as a workaround to [**`Regenerate the CSS & Data`**](https://elementor.com/help/tools/) in your **`Dashboard`** > **`Elementor`** > **`Tools`** > **`General`** for the time being.

If you find that inconvenient I found [this neat tutorial from Imran Siddiq](https://www.youtube.com/watch?v=Atbak1MS-SM) that might help. I haven't tested it myself, but it doesn't hurt to try it out (on a staging site first).

We appreciate your patience and understanding as we work on fixing this issue for good. 🙏 

Kind regards

**Comment by nicholaszein** [comment]
Hello everyone,

I just wanted to give you an update on this issue.

### 👷 We are still working on a solution. 

We have made progress, however, solving this issue affects multiple components of the builder, and we want to both solve all issues and avoid having to revert the solution back like we had to do before due to conflicts with other components.

We ask for patience as we work to address this and [other issues](https://github.com/elementor/elementor/issues/19394) that are related.

Thanks for understanding! 🙏

**Comment by nicholaszein** [comment]
Hi @VadymPogorelov, and others!

We just [released a new patch](https://github.com/orgs/elementor/discussions/23925#:~:text=Responsive%20Template%20Designs) for this issue in **version 3.17-beta**, and would love to know about your experience with it!

How is it working for you?
Did you find any issues when editing templates in responsive modes?
Is it working correctly even with cache turned on?

We want to make sure we are covering all bases here to avoid further conflicts, and we would really appreciate your feedback!

As always, please remember to make a backup before updating, and if possible test on a staging environment.

Let us know how is it working for you. 🙏

**Comment by nicholaszein** [comment]
Hello there @VadymPogorelov!

We have great news! 🙌 

### 📢 We're happy to announce that the issue you raised was resolved in Elementor v3.17! 🥳 

✅ Feel free to check it out and update your plugin to the new version!

> _Check the changelog for more information:_
> [Changelog of Elementor and Elementor Pro](https://elementor.com/help/changelog-of-elementor/)

Cheers 🥂
